sklearn/tests/test_props.py

```diff
@@ -0,0 +1,729 @@
+        assert_request_is_empty(requests)
+
+    # but one can exclude a method
+    assert_request_is_empty(requests, exclude="fit")
+
+    requests.score.add_request(prop="carrot", alias=RequestType.REQUESTED)
+    with pytest.raises(AssertionError):
+        # excluding `fit` is not enough
+        assert_request_is_empty(requests, exclude="fit")
+
+    # and excluding both fit and score would avoid an exception
+    assert_request_is_empty(requests, exclude=["fit", "score"])
+
+    # test if a router is empty
+    assert_request_is_empty(
+        MetadataRouter(owner="test")
+        .add_self(WeightedMetaRegressor(estimator=None))
+        .add(method_mapping="fit", estimator=RegressorMetadata())
+    )
+
+
+def test_default_requests():
+    class OddEstimator(BaseEstimator):
+        __metadata_request__fit = {
+            # set a different default request
+            "sample_weight": RequestType.REQUESTED
+        }  # type: ignore
+
+    odd_request = get_routing_for_object(OddEstimator())
+    assert odd_request.fit.requests == {"sample_weight": RequestType.REQUESTED}
+
+    # check other test estimators
+    assert not len(get_routing_for_object(ClassifierNoMetadata()).fit.requests)
+    assert_request_is_empty(ClassifierNoMetadata().get_metadata_routing())
+
+    trs_request = get_routing_for_object(TransformerMetadata())
+    assert trs_request.fit.requests == {
+        "sample_weight": RequestType(None),
+        "brand": RequestType(None),
+    }
+    assert trs_request.transform.requests == {
+        "sample_weight": RequestType(None),
+    }
+    assert_request_is_empty(trs_request)
+
+    est_request = get_routing_for_object(ClassifierFitMetadata())
+    assert est_request.fit.requests == {
+        "sample_weight": RequestType(None),
+        "brand": RequestType(None),
+    }
+    assert_request_is_empty(est_request)
+
+
+def test_simple_metadata_routing():
+    # Tests that metadata is properly routed
+
+    # The underlying estimator doesn't accept or request metadata
+    clf = SimpleMetaClassifier(estimator=ClassifierNoMetadata())
+    clf.fit(X, y)
+
+    # Meta-estimator consumes sample_weight, but doesn't forward it to the underlying
+    # estimator
+    clf = SimpleMetaClassifier(estimator=ClassifierNoMetadata())
+    clf.fit(X, y, sample_weight=my_weights)
+
+    # If the estimator accepts the metadata but doesn't explicitly say it doesn't
+    # need it, there's an error
+    clf = SimpleMetaClassifier(estimator=ClassifierFitMetadata())
+    with pytest.raises(
+        ValueError,
+        match=(
+            "sample_weight is passed but is not explicitly set as requested or not for"
+            " ClassifierFitMetadata.fit"
+        ),
+    ):
+        clf.fit(X, y, sample_weight=my_weights)
+
+    # Explicitly saying the estimator doesn't need it, makes the error go away,
+    # but if a metadata is passed which is not requested by any object/estimator,
+    # there will be still an error
```

Comment:
```suggestion
    # because in this case `SimpleMetaClassifier` consumes `sample_weight`
    # (whereas if there was no consumer of sample_weight, passing it would
    # result in an error.
```